### ISSUE TITLE
Windows Support (Vote)

### ISSUE THREAD

**Issue by timcharper** [sporkrb/spork#3]
Vote for this if you'd like to see Spork run on windows

**Comment by hoenth** [comment]
+1

**Comment by nzifnab** [comment]
Yes please :<

**Comment by jmonteiro** [comment]
+1

This really a needed thing to me, since I need to test my features (Cucumber) with Watir/IE :-(

**Comment by jcfischer** [comment]
+1 - we need this too for a project. I can help (hopefully) with the coding

**Comment by rdp** [comment]
+1
The kicker really is that with windows I think you'll need to forward all your stdin/stderr output across TCP sockets back to the host.  But it should work :)

**Comment by timcharper** [comment]
This is actually already done (stdin / sterr output over tcp sockets).  DRb does some magic to let you do this. (referring to Roger's comment)

I'm really excited to get started on this.  I've been very swamped with my preparations for #LSRC though (you have no idea how much work it is to prepare a measly 6 hours worth of training).  So... excited to get back to this when I get back.

**Comment by rdp** [comment]
Oh nice.  I suppose for now windows users can use cygwin--that doesn't help me too much since I use normal Test::Unit tests, but hey, it's one step closer to coding bliss :)
-r

**Comment by kriswill** [comment]
+1 working in windows would be awesome, feature runs are slooow - might have to resort to spec_server for now

**Comment by bhavinkamani** [comment]
+1

**Comment by timcharper** [comment]
Getting much closer. The RunStrategy extraction has been completed, so now one would just need to implement the Magazine RunStrategy and it'll be done!

**Comment by jcfischer** [comment]
I'll be back at the project next week (although we have switched to JRuby completely) -- and I'll take a look if I can work on this

**Comment by timcharper** [comment]
ok - jruby is on the radar as well. Can be supported via forking strategy using FFI on POSIX systems, and windows sytem via magazine strategy.

**Comment by grosser** [comment]
Just a sidenote: If anyone has experience with forking for windows/jruby, the [parallel](http://github.com/grosser/parallel) and with it the [parallel_specs](http://github.com/grosser/parallel_specs) projects are in need of windows/jruby forking too...

**Comment by pbalduino** [comment]
plz

**Comment by timcharper** [comment]
@grosser - as far as I know, ruby on windows doesn't support Kernel.fork. You will need to fire up separate processes, and communicate with them.

**Comment by rkoloski** [comment]
It'd be a huge plus for me and my team as well. 

Is it correct to assume that because Kernel.fork is implemented in Ruby, that running Cygwin won't help if Cygwin is pointing to a Win32 Ruby interpreter?

Let me know if/how I can help.

**Comment by cwoodward** [comment]
Another vote to get this working.

**Comment by timcharper** [comment]
Please see the wiki page: http://wiki.github.com/timcharper/spork/windowssupport

I've outlined what's needed to do, and I've architected spork so this will be possible.

If you're interested in implementing it, I'm happy to offer assistance.

Tim

**Comment by jsuchal** [comment]
+1

**Comment by Milhouse** [comment]
+1 Using Windows at work. Hate slow tests :)

**Comment by rdp** [comment]
If anybody's adventurous they could try my fork
http://github.com/rdp/spork
which theoretically works with windows (and give me feedback/feature requests...)
-r

**Comment by markj9** [comment]
+1

**Comment by Milhouse** [comment]
I added a fork (http://wiki.github.com/Milhouse/spork/) which adds a preliminary "magazine" strategy for Windows. It's certainly a beta, but it does work with both Rspec and Cucumber so far.

**Comment by timcharper** [comment]
Milhouse, Roger Pack, this is awesome! Way to go guys!

**Comment by Zloy** [comment]
+1

**Comment by timcharper** [comment]
merged in!

**Comment by rdp** [comment]
I think you could use popen and avoid the MRI dependency (?)

**Comment by timcharper** [comment]
that's a great idea, I was wondering about that.  Milhouse, what do you think?

**Comment by jsuchal** [comment]
Hmm, the newest gemspec file does not include magazine strategy files, please regenerate it a and push a new version to gemcutter since this is unusable at the moment. Anyways I was able to rebuild the gem locally and seems to work well. Thanks.

**Comment by timcharper** [comment]
Paciencia!

:)

It's "merged in", not released yet.

**Comment by jsuchal** [comment]
Sorry for all the impatient disturbance.

**Comment by Milhouse** [comment]
It might be a way to let JRuby join the party. I'd seen popen (and it's cousins popen3, popen4) at http://devver.net/blog/2009/10/ruby-subprocesses-part_3/   I wasn't sure it would work, so I stuck with the win-32/process library. Also, I had to change the current working directory (cwd). I'm not sure if popen has this as an option. I'll take a look, but it would early next week before I have time. Thanks again, Roger.

**Comment by Milhouse** [comment]
I'm not an expert on processes and pipes. The reason I used the win32-process gem was to create a new process that returned a PID and continued on it's merry way independent of the parent process. popen and related apis seem to expect the parent process to wait for the child to finish. There may be a way to do what I need, but it's not clear to me.

There _is_ a new call in Ruby 1.9 that does this. Process.spawn. It fires off a new process and returns the PID. It allows the working directory to be set. It works in the edge JRuby 1.5 using Ruby 1.9.2.   The PTY.spawn in Ruby 1.8 has similar functionality, but doesn't work in Windows.  I'm going to experiment to see how this might work.

**Comment by rdp** [comment]
you can probably just spawn off a new thread to wait on the popen thread.

**Comment by jetsgit** [comment]
+1

Would be very handy

jet

**Comment by rguerrero** [comment]
+1

thank!!

**Comment by rdp** [comment]
can we close this?

**Comment by timcharper** [comment]
I think it is closed. People keep on commenting on it, however :S.  okay people: it's closed! It's been done!  No more commenting!

We should probably open up a new ticket for Java support.

**Comment by rguerrero** [comment]
but exist a version for windows?

**Comment by timcharper** [comment]
yes! :)

**Comment by rguerrero** [comment]
where i can download?

**Comment by timcharper** [comment]
gem install spork.

**Comment by rguerrero** [comment]
when i do that show me this error: 
Error installing spork:
ERROR: Failed to build gem native extension.

**Comment by timcharper** [comment]
can you post the make.out file? full screen shot? any and all details will be helpful.  Use jing to record and share a quick screencast, if possible.

**Comment by rguerrero** [comment]
the gem_make.out:

"H:/Rodrigo Guerrero/RubyApp/Ruby187/bin/ruby.exe" mkrf_conf.rb
Actually, there aren't any native extensions. I'm just dynamically installing dependencies based off of your operating system
installing windows dependencies

"H:/Rodrigo Guerrero/RubyApp/Ruby187/bin/ruby.exe" -rubygems H:/Rodrigo Guerrero/RubyApp/Ruby187/lib/ruby/gems/1.8/gems/rake-0.8.7/bin/rake RUBYARCHDIR="C:/Documents and Settings/Usuario/.bundle/ruby/1.8/gems/spork-0.8.4/lib" RUBYLIBDIR="C:/Documents and Settings/Usuario/.bundle/ruby/1.8/gems/spork-0.8.4/lib"
H:/Rodrigo Guerrero/RubyApp/Ruby187/bin/ruby.exe: No such file or directory -- H:/Rodrigo (LoadError)

**Comment by rdp** [comment]
ahh. I think that is something of a bug in rubygems (doesn't handle spaces in the dir name).  I'll report it there if there are no objections to me doing so.

**Comment by rguerrero** [comment]
But i do without space for test:
"H:/Rodrig~1/RubyApp/Ruby187/bin/ruby.exe" -rubygems H:/Rodrig~1/RubyApp/Ruby187/lib/ruby/gems/1.8/gems/rake-0.8.7/bin/rake RUBYARCHDIR="C:/Docume~1/Usuario/.bundle/ruby/1.8/gems/spork-0.8.4/lib" RUBYLIBDIR="C:/Documents and Settings/Usuario/.bundle/ruby/1.8/gems/spork-0.8.4/lib"

but show me other error:
No examples matching ./spec/*_/__spec.rb could be found

**Comment by rdp** [comment]
that might be a real bug...anybody know where the root cause of that one might be found?

**Comment by rdp** [comment]
installs for me with 1.8.7 from http://rubyinstaller.org/downloads/

<pre>
D:\installs\a space\ruby-1.8.7-p249-i386-mingw32-1\ruby-1.8.7-p249-i386-mingw32\bin>gem install spork
Building native extensions.  This could take a while...
Successfully installed spork-0.8.4
1 gem installed
because of new gem installation recreating all faster_rubygems caches
faster_rubygems: creating caches
..
</pre>

Not sure about the other one though...

**Comment by rguerrero** [comment]
I have install rails 3.0.0.beta.4

**Comment by rdp** [comment]
upload your files somewhere and I can test it

**Comment by rguerrero** [comment]
give me a mail where send you.

Thank

**Comment by rdp** [comment]
rogerdpack@gmail.com
